roaring/container_stash.go

```diff
@@ -0,0 +1,263 @@
+func (c *Container) array() []uint16 {
+	if roaringParanoia {
+		if c.typ != containerArray {
+			panic("attempt to read non-array's array")
+		}
+	}
+	return *(*[]uint16)(unsafe.Pointer(&reflect.SliceHeader{Data: uintptr(unsafe.Pointer(c.pointer)), Len: int(c.len), Cap: int(c.cap)}))
+}
+
+// setArray stores a set of uint16s as data.
+func (c *Container) setArray(array []uint16) {
+	if roaringParanoia {
+		if c.typ != containerArray {
+			panic("attempt to write non-array's array")
+		}
+	}
+	// no array: start with our default 5-value array
+	if array == nil {
+		c.pointer, c.len, c.cap = (*uint16)(unsafe.Pointer(&c.data[0])), 0, stashedArraySize
+		return
+	}
+	h := (*reflect.SliceHeader)(unsafe.Pointer(&array))
+	if h.Data == uintptr(unsafe.Pointer(c.pointer)) {
+		// nothing to do but update length
+		c.len = int32(h.Len)
+		return
+	}
+	// array we can fit in data store:
+	if len(array) <= stashedArraySize {
+		copy(c.data[:stashedArraySize], array)
+		c.pointer, c.len, c.cap = (*uint16)(unsafe.Pointer(&c.data[0])), int32(len(array)), stashedArraySize
+		c.mapped = false // this is no longer using a hypothetical mmapped input array
+		return
+	}
+	c.pointer, c.len, c.cap = (*uint16)(unsafe.Pointer(h.Data)), int32(h.Len), int32(h.Cap)
+	runtime.KeepAlive(&array)
+}
+
+// bitmap yields the data viewed as a slice of uint64s holding bits.
+func (c *Container) bitmap() []uint64 {
+	if roaringParanoia {
+		if c.typ != containerBitmap {
+			panic("attempt to read non-bitmap's bitmap")
+		}
+	}
+	return *(*[]uint64)(unsafe.Pointer(&reflect.SliceHeader{Data: uintptr(unsafe.Pointer(c.pointer)), Len: int(c.len), Cap: int(c.cap)}))
+}
+
+// setBitmap stores a set of uint64s as data.
+func (c *Container) setBitmap(bitmap []uint64) {
+	if roaringParanoia {
+		if c.typ != containerBitmap {
+			panic("attempt to write non-bitmap's bitmap")
+		}
+	}
+	h := (*reflect.SliceHeader)(unsafe.Pointer(&bitmap))
+	c.pointer, c.len, c.cap = (*uint16)(unsafe.Pointer(h.Data)), int32(h.Len), int32(h.Cap)
+	runtime.KeepAlive(&bitmap)
+}
+
+// runs yields the data viewed as a slice of intervals.
+func (c *Container) runs() []interval16 {
+	if roaringParanoia {
+		if c.typ != containerRun {
+			panic("attempt to read non-run's runs")
+		}
+	}
+	return *(*[]interval16)(unsafe.Pointer(&reflect.SliceHeader{Data: uintptr(unsafe.Pointer(c.pointer)), Len: int(c.len), Cap: int(c.cap)}))
+}
+
+// setRuns stores a set of intervals as data.
+func (c *Container) setRuns(runs []interval16) {
+	if roaringParanoia {
+		if c.typ != containerRun {
+			panic("attempt to write non-run's runs")
+		}
+	}
+	// no array: start with our default 2-value array
+	if runs == nil {
+		c.pointer, c.len, c.cap = (*uint16)(unsafe.Pointer(&c.data[0])), 0, stashedRunSize
+		return
+	}
+	h := (*reflect.SliceHeader)(unsafe.Pointer(&runs))
+	if h.Data == uintptr(unsafe.Pointer(c.pointer)) {
+		// nothing to do but update length
+		c.len = int32(h.Len)
```

Comment:
I guess it's impossible that `cap` would change without the actual pointer changing?

Reply:
it's impossible that it would *correctly* change without the actual pointer changing. i believe.

... i will ask #darkarts on gopher slack, though.

Reply:
I'm out of my comfort zone, but is it possible for something like `realloc` to make an allocation larger without moving it? I'm reading the docs but it's still unclear to me. Also don't know if there's something higher level (within Go) that would prevent this.

Reply:
So far as I know, allocations have a fixed size, because they're in size spans based on their size, and the cap should always have reflected the actual size of the underlying allocation in the first place. I think.